File: sdk/typescript/src/groupStep.ts
import { BlockStep } from './blockStep'
import { CommandStep } from './commandStep'
import { InputStep } from './inputStep'
import { TriggerStep } from './triggerStep'
import { WaitStep } from './waitStep'
import { DependsOn, Notify, NotifyEnum } from './types'

export interface GroupStep {
    group: string;
    steps: Array<BlockStep | CommandStep | InputStep | TriggerStep | WaitStep>;
    allow_dependency_failure?: boolean;
    depends_on?: Array<DependsOn | string> | null | string;
    if?: string;
    key?: string;
    label?: string;
    notify?: Array<Notify | NotifyEnum>;
    skip?: boolean | string;
}

File: sdk/csharp/src/Buildkite.Sdk/GroupStep.cs
using System.Collections.Generic;

namespace Buildkite.Sdk
{
    public class GroupStep : PipelineStep
    {
        public string? Group { get; set; }
        public List<PipelineStep>? Steps { get; set; }
        public List<Notify>? Notify { get; set; }

        public GroupStep()
        {
        }

        public GroupStep(string group)
        {
            Group = group;
            Steps = new List<PipelineStep>();
        }

        public GroupStep(string label, string group)
        {
            Label = label;
            Group = group;
            Steps = new List<PipelineStep>();
        }

        public GroupStep AddStep(PipelineStep step)
        {
            Steps ??= new List<PipelineStep>();
            Steps.Add(step);
            return this;
        }
    }
}